File: cpp/slidingWindow/min_window.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

string minWindow(string s, string t) {

    unordered_map<char, int> freq, window;

    int left = 0;
    int match = 0;
    string ans = "";

    for (auto c : t) freq[c]++;

    for (int right = 0; right < s.length(); right++) {
        window[s[right]]++;
        if (freq.count(s[right]) && window[s[right]] == freq[s[right]]) match++;
        while (match == freq.size()) {
            string current = s.substr(left, right - left + 1);
            if (ans.empty() || current.length() < ans.length()) ans = current;
            if (freq.count(s[left]) && window[s[left]] == freq[s[left]]) match--;
            window[s[left]]--;
            left++;
        }
    }

    return ans;
}

int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}

File: python/slidingWindow/minWindow.py
import sys
import threading
from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        freq = defaultdict(int)
        window = defaultdict(int)
        left = 0
        match = 0
        ans = ""
        for c in t:
            freq[c] += 1

        for right in range(len(s)):
            window[s[right]] += 1
            if s[right] in freq and window[s[right]] == freq[s[right]]:
                match += 1
            while match == len(freq):
                tmp = s[left : right + 1]
                if not ans or len(tmp) < len(ans):
                    ans = tmp
                if s[left] in freq and window[s[left]] == freq[s[left]]:
                    match -= 1
                window[s[left]] -= 1
                left += 1

        return ans


def main():
    s = Solution()
    data = [1, 2, 3]


if __name__ == "__main__":
    threading.Thread(target=main).start()
